File: NeetCode/2-D_Dynamic_Programming/Coin_Change_II/coin__change__i_i.py
from typing import Optional, List


class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        dp = [0] * (amount + 1)
        dp[0] = 1

        for coin in coins:
            for index in range(coin, amount + 1):
                dp[index] = dp[index] + dp[index - coin]
        return dp[amount]


if __name__ == "__main__":
    sol = Solution()
    amount = 7
    coins = [2, 4]
    print(sol.change(amount, coins))
    print("Running Solution...")

File: NeetCode/2-D_Dynamic_Programming/Coin_Change_II/Coin_Change_II.cpp
// Coin_Change_II.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int change(int amount, vector<int> &coins)
    {
        vector<int> dp(0, amount + 1);
        dp[0] = 1;

        for (int coin : coins)
        {
            for (int index = coin; index < amount + 1; index++)
            {
                dp[index] = dp[index] + dp[index - coin];
            }
        }
        return dp[amount];
    }
};
